homeassistant/components/history/websocket_api.py

```diff
@@ -297,22 +301,31 @@ async def _async_send_historical_states(
     return last_time_dt if last_time_ts != 0 else None
 
 
-def _history_compressed_state(state: State, no_attributes: bool) -> dict[str, Any]:
-    """Convert a state to a compressed state."""
-    comp_state: dict[str, Any] = {COMPRESSED_STATE_STATE: state.state}
+def _history_compressed_state(
+    state: State, no_attributes: bool
+) -> HistoryCompressedState | HistoryMinimalCompressedState:
+    """Convert a state to a compressed state.
+
+    Returns HistoryMinimalCompressedState if no_attributes is True.
+    """
+    comp_state: dict[str, Any] = {
+        COMPRESSED_STATE_STATE: state.state,
+        COMPRESSED_STATE_LAST_UPDATED: state.last_updated_timestamp,
+    }
     if not no_attributes or state.domain in history.NEED_ATTRIBUTE_DOMAINS:
         comp_state[COMPRESSED_STATE_ATTRIBUTES] = state.attributes
-    comp_state[COMPRESSED_STATE_LAST_UPDATED] = state.last_updated_timestamp
     if state.last_changed != state.last_updated:
         comp_state[COMPRESSED_STATE_LAST_CHANGED] = state.last_changed_timestamp
-    return comp_state
+    return comp_state  # type: ignore[return-value]
 
 
 def _events_to_compressed_states(
     events: Iterable[Event], no_attributes: bool
-) -> MutableMapping[str, list[dict[str, Any]]]:
+) -> MutableMapping[str, list[HistoryCompressedState | HistoryMinimalCompressedState]]:
```

Comment:
The return typing is not correct, it should be `MutableMapping[str, list[HistoryCompressedState] | list[HistoryMinimalCompressedState]]`